lib/worker/output.go

```diff
@@ -0,0 +1,220 @@
+	// the readers slice is initialised on the constructor and should never
+	// change afterwards
+	readers []io.Reader
+	// combinedBuffer keeps the combined output, this is used to forward to
+	// new listeners that came in late
+	combinedBuffer []byte
+	// messages is a channel where new output is delivered
+	messages chan []byte
+	// errors is a channel to forward output reading errors to listeners
+	errors chan error
+	// done is a channel to notify listeners that all output has been read
+	done chan struct{}
+	mu   *sync.Mutex
+}
+
+// ProcessOutputEntry is the struct sent to clients' channels
+// streaming output
+type ProcessOutputEntry struct {
+	Content    []byte
+	ReceivedAt time.Time
+}
+
+// Listener represents an output listener that the output handler will
+// send updates to.
+type Listener struct {
+	outputChan chan ProcessOutputEntry
+	errorChan  chan error
+}
+
+// NewOutputHandler returns a new OutputHandler struct from the provided readers.
+func NewOutputHandler(rc ...io.Reader) (*OutputHandler, error) {
+	if len(rc) == 0 {
+		return nil, fmt.Errorf("must provide at least one io.Reader")
+	}
+	o := &OutputHandler{
+		readers:  rc,
+		mu:       &sync.Mutex{},
+		done:     make(chan struct{}),
+		messages: make(chan []byte),
+		errors:   make(chan error),
+	}
+	go o.handleOutput()
+	go o.handleBroadcast()
+	go o.handleErrors()
+	return o, nil
+}
+
+// Stream returns a channel to receive output up to the
+// present time and stream real time. The channel closes when
+// there the process the output is coming from is finished.
+// Any errors encountered during output reading are sent through
+// the errors channel.
+func (o *OutputHandler) Stream() (<-chan ProcessOutputEntry, <-chan error) {
+	return o.addListener()
+}
+
+// addListener registers a new listener in the output handler and returns
+// channels that get all previous output and stream new output.
+// The whole operation is behind a lock as this cannot happen at the same
+// time as listeners are sent new output or the buffer is being updated.
+func (o *OutputHandler) addListener() (chan ProcessOutputEntry, chan error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	// we are going to be sending the existing output entry which is at
+	// most one item
+	output := make(chan ProcessOutputEntry, 1)
+	errChan := make(chan error)
+
+	// catch up on existing output but only send it if there
+	// is anything to send.
+	if len(o.combinedBuffer) > 0 {
+		output <- ProcessOutputEntry{Content: o.combinedBuffer}
+	}
+	select {
+	case <-o.done:
+		close(output)
+	default:
+		o.listeners = append(o.listeners, &Listener{
+			outputChan: output,
+			errorChan:  errChan,
+		})
+	}
+	return output, errChan
+}
+
+func (o *OutputHandler) updateCombinedBuffer(b []byte) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.combinedBuffer = append(o.combinedBuffer, b...)
+}
+
+func (o *OutputHandler) iterListeners(routine func(*Listener)) {
+	o.mu.Lock()
```

Comment:
Do you really need to hold the lock for the entire time this is running? Can you just lock the field read and then iterate over it outside of the lock?

Reply:
Yeah I think you are right, I don't. My idea with iterListeners was to prevent both the buffers and the listeners being tinkered with at the same time but this is problematic since we are using iterListeners for channel operations. I applied this change (see also https://github.com/trivelle/worker/pull/2#discussion_r859602188)